File: native/ios/WatermelonDB/DatabaseDriver.swift
import Foundation

class DatabaseDriver {
    typealias SchemaVersion = Int
    typealias Schema = (version: SchemaVersion, sql: Database.SQL)
    typealias MigrationSet = (from: SchemaVersion, to: SchemaVersion, sql: Database.SQL)

    struct SchemaNeededError: Error { }
    struct MigrationNeededError: Error {
        let databaseVersion: SchemaVersion
    }

    let database: Database

    convenience init(dbName: String, schemaVersion: SchemaVersion) throws {
        self.init(dbName: dbName)

        switch isCompatible(withVersion: schemaVersion) {
        case .compatible: break
        case .needsSetup:
            throw SchemaNeededError()
        case .needsMigration(fromVersion: let dbVersion):
            throw MigrationNeededError(databaseVersion: dbVersion)
        }
    }

    convenience init(dbName: String, setUpWithSchema schema: Schema) {
        self.init(dbName: dbName)

        do {
            try unsafeResetDatabase(schema: schema)
        } catch {
            fatalError("Error while setting up the database: \(error)")
        }
    }

    convenience init(dbName: String, setUpWithMigrations migrations: MigrationSet) throws {
        self.init(dbName: dbName)
        try migrate(with: migrations)
    }

    private init(dbName: String) {
        self.database = Database(path: getPath(dbName: dbName))
    }

    func find(table: Database.TableName, id: RecordId) throws -> Any? {
        guard !isCached(table, id) else {
            return id
        }

        let results = try database.queryRaw("select * from `\(table)` where id == ? limit 1", [id])

        guard let record = results.next() else {
            return nil
        }

        markAsCached(table, id)
        return record.resultDictionary!
    }

    func cachedQuery(table: Database.TableName, query: Database.SQL) throws -> [Any] {
        return try database.queryRaw(query).map { row in
            let id = row.string(forColumn: "id")!

            if isCached(table, id) {
                return id
            } else {
                markAsCached(table, id)
                return row.resultDictionary!
            }
        }
    }

    func count(_ query: Database.SQL) throws -> Int {
        return try database.count(query)
    }

    enum Operation {
        case copy(table: Database.TableName, attachPath: String)
        case execute(table: Database.TableName, query: Database.SQL, args: Database.QueryArgs)
        case create(table: Database.TableName, id: RecordId, query: Database.SQL, args: Database.QueryArgs)
        case destroyPermanently(table: Database.TableName, id: RecordId)
        case markAsDeleted(table: Database.TableName, id: RecordId)
        // case destroyDeletedRecords(table: Database.TableName, records: [RecordId])
        // case setLocal(key: String, value: String)
        // case removeLocal(key: String)
    }

    func copyTables(_ tables: [String], srcDB: String) throws {
        try database.execute("ATTACH DATABASE '\(srcDB)' as 'other'")

        try database.inTransaction {
            for table in tables {
                try database.execute("INSERT OR IGNORE  INTO \(table) SELECT * FROM other.\(table)")
            }
        }

        try database.execute("DETACH DATABASE 'other'")
    }

    func batch(_ operations: [Operation]) throws {
        var newIds: [(Database.TableName, RecordId)] = []
        var removedIds: [(Database.TableName, RecordId)] = []

        try database.inTransaction {
            for operation in operations {
                switch operation {
                case .execute(table: _, query: let query, args: let args):
                    try database.execute(query, args)

                case .create(table: let table, id: let id, query: let query, args: let args):
                    try database.execute(query, args)
                    newIds.append((table, id))

                case .markAsDeleted(table: let table, id: let id):
                    try database.execute("update `\(table)` set _status='deleted' where id == ?", [id])
                    removedIds.append((table, id))

                case .destroyPermanently(table: let table, id: let id):
                    // TODO: What's the behavior if nothing got deleted?
                    try database.execute("delete from `\(table)` where id == ?", [id])
                    removedIds.append((table, id))

                case .copy(table: let table, attachPath: let attachPath):
                    try database.executeStatements("ATTACH DATABASE '\(attachPath)' as 'other'; INSERT INTO \(table) SELECT * FROM other.\(table)")

                }
            }
        }

        for (table, id) in newIds {
            markAsCached(table, id)
        }

        for (table, id) in removedIds {
            removeFromCache(table, id)
        }
    }
    
    func getDeletedRecords(table: Database.TableName) throws -> [RecordId] {
        return try database.queryRaw("select id from `\(table)` where _status='deleted'").map { row in
            row.string(forColumn: "id")!
        }
    }

    func destroyDeletedRecords(table: Database.TableName, records: [RecordId]) throws {
        // TODO: What's the behavior if record doesn't exist or isn't actually deleted?
        let recordPlaceholders = records.map { _ in "?" }.joined(separator: ",")
        try database.execute("delete from `\(table)` where id in (\(recordPlaceholders))", records)
    }

// MARK: - LocalStorage

    func getLocal(key: String) throws -> String? {
        let results = try database.queryRaw("select `value` from `local_storage` where `key` = ?", [key])

        guard let record = results.next() else {
            return nil
        }

        return record.string(forColumn: "value")!
    }

    func setLocal(key: String, value: String) throws {
        return try database.execute("insert or replace into `local_storage` (key, value) values (?, ?)", [key, value])
    }

    func removeLocal(key: String) throws {
        return try database.execute("delete from `local_storage` where `key` == ?", [key])
    }

// MARK: - Record caching

    typealias RecordId = String

    // Rewritten to use good ol' mutable Objective C for performance
    // The swifty implementation in debug took >100s to execute on a 65K batch. This: 6ms. Yes. Really.
    private var cachedRecords: NSMutableDictionary /* [TableName: Set<RecordId>] */ = NSMutableDictionary()

    func isCached(_ table: Database.TableName, _ id: RecordId) -> Bool {
        if let set = cachedRecords[table] as? NSSet {
            return set.contains(id)
        }
        return false
    }

    private func markAsCached(_ table: Database.TableName, _ id: RecordId) {
        var cachedSet: NSMutableSet
        if let set = cachedRecords[table] as? NSMutableSet {
            cachedSet = set
        } else {
            cachedSet = NSMutableSet()
            cachedRecords[table] = cachedSet
        }
        cachedSet.add(id)
    }

    private func removeFromCache(_ table: Database.TableName, _ id: RecordId) {
        if let set = cachedRecords[table] as? NSMutableSet {
            set.remove(id)
        }
    }

// MARK: - Other private details

    private enum SchemaCompatibility {
        case compatible
        case needsSetup
        case needsMigration(fromVersion: SchemaVersion)
    }

    private func isCompatible(withVersion schemaVersion: SchemaVersion) -> SchemaCompatibility {
        let databaseVersion = database.userVersion

        switch databaseVersion {
        case schemaVersion: return .compatible
        case 0: return .needsSetup
        case (1..<schemaVersion): return .needsMigration(fromVersion: databaseVersion)
        default:
            consoleLog("Database has newer version (\(databaseVersion)) than what the " +
                "app supports (\(schemaVersion)). Will reset database.")
            return .needsSetup
        }
    }

    func unsafeResetDatabase(schema: Schema) throws {
        try database.unsafeDestroyEverything()
        cachedRecords = [:]

        try setUpSchema(schema: schema)
    }

    private func setUpSchema(schema: Schema) throws {
        try database.inTransaction {
            try database.executeStatements(schema.sql + localStorageSchema)
            database.userVersion = schema.version
        }
    }

    private func migrate(with migrations: MigrationSet) throws {
        precondition(
            database.userVersion == migrations.from,
            "Incompatbile migration set applied. DB: \(database.userVersion), migration: \(migrations.from)"
        )

        try database.inTransaction {
            try database.executeStatements(migrations.sql)
            database.userVersion = migrations.to
        }
    }

    private let localStorageSchema = """
        create table local_storage (
        key varchar(16) primary key not null,
        value text not null
        );

        create index local_storage_key_index on local_storage (key);
    """
}

private func getPath(dbName: String) -> String {
    // If starts with `file:` or contains `/`, it's a path!
    if dbName.starts(with: "file:") || dbName.contains("/") {
        return dbName
    } else {
        // swiftlint:disable:next force_try
        return try! FileManager.default
            .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false)
            .appendingPathComponent("\(dbName).db")
            .path
    }
}

File: src/adapters/sqlite/makeDispatcher/node/DatabaseDriver.js
// @flow

import Database from './Database'

function fixArgs(args: any): any {
  return Object.keys(args).reduce((acc, argName) => {
    if (typeof acc[argName] === 'boolean') {
      acc[argName] = acc[argName] ? 1 : 0
    }
    return acc
  }, args)
}

type Migrations = { from: number, to: number, sql: string }

class MigrationNeededError extends Error {
  databaseVersion: number

  type: string

  constructor(databaseVersion): void {
    super('MigrationNeededError')
    this.databaseVersion = databaseVersion
    this.type = 'MigrationNeededError'
    this.message = 'MigrationNeededError'
  }
}

class SchemaNeededError extends Error {
  type: string

  constructor(): void {
    super('SchemaNeededError')
    this.type = 'SchemaNeededError'
    this.message = 'SchemaNeededError'
  }
}

export function getPath(dbName: string): string {
  if (dbName === ':memory:' || dbName === 'file::memory:') {
    return dbName
  }

  let path =
    dbName.startsWith('/') || dbName.startsWith('file:') ? dbName : `${process.cwd()}/${dbName}`
  if (path.indexOf('.db') === -1) {
    if (path.indexOf('?') >= 0) {
      const index = path.indexOf('?')
      path = `${path.substring(0, index)}.db${path.substring(index)}`
    } else {
      path = `${path}.db`
    }
  }

  return path
}

class DatabaseDriver {
  static sharedMemoryConnections: {...} = {}

  database: Database

  cachedRecords: any = {}

  initialize: ((dbName: string, schemaVersion: number) => void) = (dbName: string, schemaVersion: number) => {
    this.init(dbName)
    this.isCompatible(schemaVersion)
  }

  setUpWithSchema: ((dbName: string, schema: string, schemaVersion: number) => void) = (dbName: string, schema: string, schemaVersion: number) => {
    this.init(dbName)
    this.unsafeResetDatabase({ version: schemaVersion, sql: schema })
    this.isCompatible(schemaVersion)
  }

  setUpWithMigrations: ((dbName: string, migrations: Migrations) => void) = (dbName: string, migrations: Migrations) => {
    this.init(dbName)
    this.migrate(migrations)
    this.isCompatible(migrations.to)
  }

  init: ((dbName: string) => void) = (dbName: string) => {
    this.database = new Database(getPath(dbName))

    const isSharedMemory = dbName.indexOf('mode=memory') > 0 && dbName.indexOf('cache=shared') > 0
    if (isSharedMemory) {
      if (!DatabaseDriver.sharedMemoryConnections[dbName]) {
        DatabaseDriver.sharedMemoryConnections[dbName] = this.database
      }
      this.database = DatabaseDriver.sharedMemoryConnections[dbName]
    }
  }

  find: ((table: string, id: string) => any | null | string) = (table: string, id: string) => {
    if (this.isCached(table, id)) {
      return id
    }

    const query = `SELECT * FROM '${table}' WHERE id == ? LIMIT 1`
    const results = this.database.queryRaw(query, [id])

    if (results.length === 0) {
      return null
    }

    this.markAsCached(table, id)
    return results[0]
  }

  cachedQuery: ((table: string, query: string) => Array<any>) = (table: string, query: string): any[] => {
    const results = this.database.queryRaw(query)
    return results.map((row: any) => {
      const id = `${row.id}`
      if (this.isCached(table, id)) {
        return id
      }
      this.markAsCached(table, id)
      return row
    })
  }

  query: ((table: string, query: string) => Array<any>) = (table: string, query: string) => this.cachedQuery(table, query)

  count: ((query: string) => number) = (query: string) => this.database.count(query)

  batch: ((operations: Array<any>) => void) = (operations: any[]) => {
    const newIds = []
    const removedIds = []

    this.database.inTransaction(() => {
      operations.forEach((operation: any[]) => {
        const [type, table, ...rest] = operation
        switch (type) {
          case 'execute': {
            const [query, args] = rest
            this.database.execute(query, fixArgs(args))
            break
          }

          case 'create': {
            const [id, query, args] = rest
            this.database.execute(query, fixArgs(args))
            newIds.push([table, id])
            break
          }

          case 'markAsDeleted': {
            const [id] = rest
            const query = `UPDATE '${table}' SET _status='deleted' WHERE id == ?`
            this.database.execute(query, [id])
            removedIds.push([table, id])
            break
          }

          case 'destroyPermanently': {
            const [id] = rest
            // TODO: What's the behavior if nothing got deleted?
            this.database.execute(`DELETE FROM '${table}' WHERE id == ?`, [id])
            removedIds.push([table, id])
            break
          }

          default: {
            throw new Error('unreachable')
          }
        }
      })
    })

    newIds.forEach(([table, id]) => {
      this.markAsCached(table, id)
    })

    removedIds.forEach(([table, id]) => {
      this.removeFromCache(table, id)
    })
  }

  getDeletedRecords: ((table: string) => Array<string>) = (table: string): string[] => {
    return this.database
      .queryRaw(`SELECT ID FROM '${table}' WHERE _status='deleted'`)
      .map(row => `${row.id}`)
  }

  destroyDeletedRecords: ((table: string, records: Array<string>) => void) = (table: string, records: string[]) => {
    const recordPlaceholders = records.map(() => '?').join(',')
    this.database.execute(`DELETE FROM '${table}' WHERE id IN (${recordPlaceholders})`, records)
  }

  // MARK: - LocalStorage

  getLocal: ((key: string) => any | null) = (key: string) => {
    const results = this.database.queryRaw('SELECT `value` FROM `local_storage` WHERE `key` = ?', [
      key,
    ])

    if (results.length > 0) {
      return results[0].value
    }

    return null
  }

  setLocal: ((key: string, value: any) => void) = (key: string, value: any) => {
    this.database.execute('INSERT OR REPLACE INTO `local_storage` (key, value) VALUES (?, ?)', [
      key,
      `${value}`,
    ])
  }

  removeLocal: ((key: string) => void) = (key: string) => {
    this.database.execute('DELETE FROM `local_storage` WHERE `key` == ?', [key])
  }

  // MARK: - Record caching

  hasCachedTable: ((table: string) => any) = (table: string) =>
    Object.prototype.hasOwnProperty.call(this.cachedRecords, table)

  isCached: ((table: string, id: string) => any | boolean) = (table: string, id: string) => {
    if (this.hasCachedTable(table)) {
      return this.cachedRecords[table].has(id)
    }
    return false
  }

  markAsCached: ((table: string, id: string) => void) = (table: string, id: string) => {
    if (!this.hasCachedTable(table)) {
      this.cachedRecords[table] = new Set()
    }
    this.cachedRecords[table].add(id)
  }

  removeFromCache: ((table: string, id: string) => void) = (table: string, id: string) => {
    if (this.hasCachedTable(table) && this.cachedRecords[table].has(id)) {
      this.cachedRecords[table].delete(id)
    }
  }

  // MARK: - Other private details

  isCompatible: ((schemaVersion: number) => void) = (schemaVersion: number) => {
    const databaseVersion = this.database.userVersion
    if (schemaVersion !== databaseVersion) {
      if (databaseVersion > 0 && databaseVersion < schemaVersion) {
        throw new MigrationNeededError(databaseVersion)
      } else {
        throw new SchemaNeededError()
      }
    }
  }

  unsafeResetDatabase: ((schema: {sql: string, version: number,...}) => void) = (schema: { sql: string, version: number }) => {
    this.database.unsafeDestroyEverything()
    this.cachedRecords = {}

    this.setUpSchema(schema)
  }

  setUpSchema: ((schema: {sql: string, version: number,...}) => void) = (schema: { sql: string, version: number }) => {
    this.database.inTransaction(() => {
      this.database.executeStatements(schema.sql + this.localStorageSchema)
      this.database.userVersion = schema.version
    })
  }

  migrate: ((migrations: Migrations) => void) = (migrations: Migrations) => {
    const databaseVersion = this.database.userVersion

    if (`${databaseVersion}` !== `${migrations.from}`) {
      throw new Error(
        `Incompatbile migration set applied. DB: ${databaseVersion}, migration: ${migrations.from}`,
      )
    }

    this.database.inTransaction(() => {
      this.database.executeStatements(migrations.sql)
      this.database.userVersion = migrations.to
    })
  }

  localStorageSchema: string = `
      create table local_storage (
      key varchar(16) primary key not null,
      value text not null
      );

      create index local_storage_key_index on local_storage (key);
      `
}

export default DatabaseDriver
